File: apps/indexer-sol/src/decode.ts
import { createHash } from "node:crypto";
import bs58 from "bs58";
import {
  CURVE_EVENTS,
  EVENT_TO_HANDLER,
  FACTORY_EVENTS,
  TREASURY_EVENTS,
  type SolanaHandlerName,
} from "@pump/solana-sdk";

export type DecodedSolanaEvent = {
  name: string;
  handler: SolanaHandlerName | null;
  dataBase64: string;
  /** Borsh-decoded fields when discriminator matched. */
  fields: Record<string, unknown> | null;
  signature: string;
  slot: number;
  programId: string;
  /** Index among Program data lines in this log batch (trade uniqueness). */
  logIndex: number;
};

type Reader = { buf: Buffer; offset: number };

function eventDiscriminator(name: string): Buffer {
  return createHash("sha256").update(`event:${name}`).digest().subarray(0, 8);
}

const DISC_TO_NAME: Map<string, string> = (() => {
  const names = [
    FACTORY_EVENTS.TokenCreated,
    CURVE_EVENTS.TokenRegistered,
    CURVE_EVENTS.TradeEvent,
    CURVE_EVENTS.FeeSplitEvent,
    CURVE_EVENTS.ReferrerSetEvent,
    CURVE_EVENTS.FeesClaimed,
    TREASURY_EVENTS.TreasuryWithdraw,
  ];
  const map = new Map<string, string>();
  for (const name of names) {
    map.set(eventDiscriminator(name).toString("hex"), name);
  }
  return map;
})();

function readPubkey(r: Reader): string {
  const slice = r.buf.subarray(r.offset, r.offset + 32);
  r.offset += 32;
  return bs58.encode(slice);
}

function readU64(r: Reader): bigint {
  const v = r.buf.readBigUInt64LE(r.offset);
  r.offset += 8;
  return v;
}

function readU8(r: Reader): number {
  const v = r.buf.readUInt8(r.offset);
  r.offset += 1;
  return v;
}

function readBool(r: Reader): boolean {
  return readU8(r) !== 0;
}

function readString(r: Reader): string {
  const len = r.buf.readUInt32LE(r.offset);
  r.offset += 4;
  const s = r.buf.subarray(r.offset, r.offset + len).toString("utf8");
  r.offset += len;
  return s;
}

function decodeFields(name: string, body: Buffer): Record<string, unknown> {
  const r: Reader = { buf: body, offset: 0 };
  switch (name) {
    case FACTORY_EVENTS.TokenCreated:
      return {
        mint: readPubkey(r),
        creator: readPubkey(r),
        name: readString(r),
        symbol: readString(r),
        uri: readString(r),
        totalSupply: readU64(r),
        virtualSolReserve: readU64(r),
        decimals: readU8(r),
      };
    case CURVE_EVENTS.TokenRegistered:
      return {
        mint: readPubkey(r),
        creator: readPubkey(r),
        totalSupply: readU64(r),
        virtualSolReserve: readU64(r),
        virtualTokenReserve: readU64(r),
      };
    case CURVE_EVENTS.TradeEvent:
      return {
        mint: readPubkey(r),
        trader: readPubkey(r),
        isBuy: readBool(r),
        solAmount: readU64(r),
        tokenAmount: readU64(r),
        feeLamports: readU64(r),
        reserveSol: readU64(r),
        soldTokens: readU64(r),
        spotPrice: readU64(r),
      };
    case CURVE_EVENTS.FeeSplitEvent:
      return {
        mint: readPubkey(r),
        creator: readPubkey(r),
        creatorFee: readU64(r),
        referrerFee: readU64(r),
        treasuryFee: readU64(r),
      };
    case CURVE_EVENTS.ReferrerSetEvent:
      return {
        trader: readPubkey(r),
        referrer: readPubkey(r),
      };
    case CURVE_EVENTS.FeesClaimed:
      return {
        owner: readPubkey(r),
        amount: readU64(r),
      };
    case TREASURY_EVENTS.TreasuryWithdraw:
      return {
        to: readPubkey(r),
        amount: readU64(r),
      };
    default:
      return {};
  }
}

/** Decode one Anchor `Program data:` base64 blob. Returns null if unknown disc. */
export function decodeProgramData(
  dataBase64: string
): { name: string; fields: Record<string, unknown> } | null {
  let buf: Buffer;
  try {
    buf = Buffer.from(dataBase64, "base64");
  } catch {
    return null;
  }
  if (buf.length < 8) return null;
  const name = DISC_TO_NAME.get(buf.subarray(0, 8).toString("hex"));
  if (!name) return null;
  try {
    return { name, fields: decodeFields(name, buf.subarray(8)) };
  } catch (err) {
    console.warn(`[indexer-sol] borsh decode failed for ${name}:`, err);
    return null;
  }
}

export function extractEventsFromLogs(input: {
  logs: string[];
  signature: string;
  slot: number;
  programId: string;
}): DecodedSolanaEvent[] {
  const out: DecodedSolanaEvent[] = [];
  let logIndex = 0;

  for (const line of input.logs) {
    const dataMatch = /^Program data:\s*(\S+)$/.exec(line);
    if (!dataMatch) continue;

    const dataBase64 = dataMatch[1];
    const decoded = decodeProgramData(dataBase64);
    if (!decoded) continue;

    const handler =
      decoded.name in EVENT_TO_HANDLER
        ? EVENT_TO_HANDLER[decoded.name as keyof typeof EVENT_TO_HANDLER]
        : null;

    out.push({
      name: decoded.name,
      handler,
      dataBase64,
      fields: decoded.fields,
      signature: input.signature,
      slot: input.slot,
      programId: input.programId,
      logIndex: logIndex++,
    });
  }

  return out;
}

/** Exported for unit tests / tooling. */
export function eventDiscHex(name: string): string {
  return eventDiscriminator(name).toString("hex");
}

File: apps/indexer-sol-go/internal/decode/decode.go
package decode

import (
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"regexp"
	"strings"

	"github.com/mr-tron/base58"
)

var programDataRe = regexp.MustCompile(`^Program data:\s*(\S+)$`)

// Event names — must match @pump/solana-sdk CURVE_EVENTS / FACTORY_EVENTS.
const (
	EventTokenCreated        = "TokenCreated"
	EventTokenRegistered     = "TokenRegistered"
	EventTradeEvent          = "TradeEvent"
	EventFeeSplitEvent       = "FeeSplitEvent"
	EventFeeSplitV2Event     = "FeeSplitV2Event"
	EventReferrerSetEvent    = "ReferrerSetEvent"
	EventCreatorFeeClaimed   = "CreatorFeeClaimed"
	EventReferrerFeeClaimed  = "ReferrerFeeClaimed"
	EventEmergencyEthSwept   = "EmergencyEthSwept"
	EventTreasuryWithdraw    = "TreasuryWithdraw"
)

var discToName map[string]string

func init() {
	names := []string{
		EventTokenCreated,
		EventTokenRegistered,
		EventTradeEvent,
		EventFeeSplitEvent,
		EventFeeSplitV2Event,
		EventReferrerSetEvent,
		EventCreatorFeeClaimed,
		EventReferrerFeeClaimed,
		EventEmergencyEthSwept,
		EventTreasuryWithdraw,
	}
	discToName = make(map[string]string, len(names))
	for _, name := range names {
		discToName[eventDiscHex(name)] = name
	}
}

func eventDiscHex(name string) string {
	sum := sha256.Sum256([]byte("event:" + name))
	return hex.EncodeToString(sum[:8])
}

// EventDiscHex exports discriminator for parity tests.
func EventDiscHex(name string) string {
	return eventDiscHex(name)
}

type Event struct {
	Name      string
	Signature string
	Slot      uint64
	ProgramID string
	LogIndex  int
	Fields    map[string]any
}

type reader struct {
	buf []byte
	off int
}

func (r *reader) readPubkey() string {
	if r.off+32 > len(r.buf) {
		panic("short pubkey")
	}
	out := base58.Encode(r.buf[r.off : r.off+32])
	r.off += 32
	return out
}

func (r *reader) readU64() uint64 {
	if r.off+8 > len(r.buf) {
		panic("short u64")
	}
	v := binary.LittleEndian.Uint64(r.buf[r.off:])
	r.off += 8
	return v
}

func (r *reader) readU32() uint32 {
	if r.off+4 > len(r.buf) {
		panic("short u32")
	}
	v := binary.LittleEndian.Uint32(r.buf[r.off:])
	r.off += 4
	return v
}

func (r *reader) readU8() uint8 {
	if r.off+1 > len(r.buf) {
		panic("short u8")
	}
	v := r.buf[r.off]
	r.off++
	return v
}

func (r *reader) readString() string {
	length := int(r.readU32())
	if r.off+length > len(r.buf) {
		panic("short string")
	}
	s := string(r.buf[r.off : r.off+length])
	r.off += length
	return s
}

func decodeFields(name string, body []byte) (map[string]any, error) {
	r := reader{buf: body}
	switch name {
	case EventTokenCreated:
		return map[string]any{
			"mint":              r.readPubkey(),
			"creator":           r.readPubkey(),
			"name":              r.readString(),
			"symbol":            r.readString(),
			"uri":               r.readString(),
			"totalSupply":       r.readU64(),
			"virtualSolReserve": r.readU64(),
			"decimals":          r.readU8(),
		}, nil
	case EventTokenRegistered:
		return map[string]any{
			"mint":                r.readPubkey(),
			"creator":             r.readPubkey(),
			"totalSupply":         r.readU64(),
			"virtualSolReserve":   r.readU64(),
			"virtualTokenReserve": r.readU64(),
		}, nil
	case EventTradeEvent:
		return map[string]any{
			"mint":         r.readPubkey(),
			"trader":       r.readPubkey(),
			"isBuy":        r.readU8() != 0,
			"solAmount":    r.readU64(),
			"tokenAmount":  r.readU64(),
			"feeLamports":  r.readU64(),
			"reserveSol":   r.readU64(),
			"soldTokens":   r.readU64(),
			"spotPrice":    r.readU64(),
		}, nil
	case EventFeeSplitEvent:
		return map[string]any{
			"mint":         r.readPubkey(),
			"creator":      r.readPubkey(),
			"creatorFee":   r.readU64(),
			"referrerFee":  r.readU64(),
			"treasuryFee":  r.readU64(),
		}, nil
	case EventFeeSplitV2Event:
		return map[string]any{
			"mint":          r.readPubkey(),
			"creator":       r.readPubkey(),
			"creatorFee":    r.readU64(),
			"referrerFee":   r.readU64(),
			"cashbackFee":   r.readU64(),
			"clanPoolFee":   r.readU64(),
			"seasonPoolFee": r.readU64(),
			"platformFee":   r.readU64(),
			"userXp":        r.readU32(),
		}, nil
	case EventReferrerSetEvent:
		return map[string]any{
			"trader":   r.readPubkey(),
			"referrer": r.readPubkey(),
		}, nil
	case EventCreatorFeeClaimed:
		return map[string]any{
			"creator": r.readPubkey(),
			"amount":  r.readU64(),
		}, nil
	case EventReferrerFeeClaimed:
		return map[string]any{
			"referrer": r.readPubkey(),
			"amount":   r.readU64(),
		}, nil
	case EventEmergencyEthSwept:
		return map[string]any{
			"to":     r.readPubkey(),
			"amount": r.readU64(),
		}, nil
	case EventTreasuryWithdraw:
		return map[string]any{
			"to":     r.readPubkey(),
			"amount": r.readU64(),
		}, nil
	default:
		return map[string]any{}, nil
	}
}

func DecodeProgramData(dataBase64 string) (string, map[string]any, error) {
	raw, err := base64.StdEncoding.DecodeString(dataBase64)
	if err != nil {
		return "", nil, err
	}
	if len(raw) < 8 {
		return "", nil, fmt.Errorf("too short")
	}
	name, ok := discToName[hex.EncodeToString(raw[:8])]
	if !ok {
		return "", nil, fmt.Errorf("unknown discriminator")
	}
	fields, err := decodeFields(name, raw[8:])
	if err != nil {
		return "", nil, err
	}
	return name, fields, nil
}

func ExtractEventsFromLogs(logs []string, signature, programID string, slot uint64) []Event {
	out := make([]Event, 0, 2)
	logIndex := 0
	for _, line := range logs {
		m := programDataRe.FindStringSubmatch(strings.TrimSpace(line))
		if len(m) != 2 {
			continue
		}
		name, fields, err := DecodeProgramData(m[1])
		if err != nil {
			continue
		}
		out = append(out, Event{
			Name:      name,
			Signature: signature,
			Slot:      slot,
			ProgramID: programID,
			LogIndex:  logIndex,
			Fields:    fields,
		})
		logIndex++
	}
	return out
}
